[WIP] [8.x] Backport http2 changes from 10.x

kjin opened the pull request:
This PR tracks an effort to backport new HTTP/2 features from 10.x. It's a work in progress, so this PR is currently more of an FYI and shouldn't really be reviewed (except if it's clear that a commit doesn't really belong).

Note that I have some intermediate commits that currently skip some tests.

cc/ @ofrobots @apapirovski @nodejs/http2

##### Checklist

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

mcollina commented:
Good work! I've tagged this as semver-minor because it adds a new feature to http as well.

jasnell commented:
Very nice! Thank you for taking this on

addaleax commented:
It actually looks like this already needs a rebase? :/

kjin commented:
@addaleax Yep, I will be doing the rebase in a bit. Note that the commits as of right now represent maybe 50% of the commits that will be part of this PR... more to come.

kjin commented:
Update: I've backported commits that have to do with HTTP/2 all the way up to August 15. Based on how many come afterward I'm thinking of just opening a new PR for remaining commits.

I created list of commits based on any commit the all matched the following criteria:
- Has not been backported to 8.x before, and associated PR does not have the `do-not-backport-v8.x` label
- Is on the branch `v10.x-staging` (as of August 15)
- Satisfies at least one of the following:
  - Touches files that match at least one of these globs: `lib/internal/http2`, `doc/api/http2.md`, `test/**/test-http2-*`, `src/node_http2*`
  - Has `http2` in the commit name
  - Is from a PR with the `http2` label
I removed commits that I believed were too broad to be applied to a PR based around just backporting http2, which resulted in (1) the commits in the PR, and (2) a smaller list of commits that I did not add to this PR but are worth noting. I grouped these commits below based on what I think should be backported together in follow-up PRs:

* This seems to be part of a "StreamBase refactor" which I wasn't sure whether it belonged in v8.x
  * e85c20b511 net,http2: merge write error handling & property names (#19734)
* Changes to `errors.js` seem independent of H2 changes
  * d28211ec3d errors: validate input arguments (#19924)
  * b6fbe16b41 errors: alter ERR_HTTP2_INVALID_CONNECTION_HEADERS (#19807)
  * f086354d3b errors: alter ERR_HTTP2_PSEUDOHEADER_NOT_ALLOWED (#19958)
* Runtime-Controlled Debugging System -- this seems fairly independent despite the change being mostly in HTTP2
  * 182c73bf7f http2: switch to new runtime-controlled debugging system (#20987)
* These are a dependent chain of commits that didn't seem to merge cleanly
  * 0b3c80ca31 http2: fix issues with aborted `respondWithFile()`s (#21561)
  * eea199bf98 test: fix http2 connection abort test (#21861)
  * c3d9000111 doc: document http2 network error behaviour (#21861)
  * 793a5bb4af test: don't fail http2 abort test if 'data' is called multiple times (#21925)

@addaleax, it seems like you are the author of the majority of these changes, so any advice you might have on how to proceed with these would be appreciated.

kjin commented:
Ping @MylesBorins @addaleax @nodejs/http2 -- Since Node 10 will be going to LTS soon, I want to make sure we have enough time to react to any requested changes/future PRs.

MylesBorins commented:
@kjin thanks for the hard work! We should plan for another 8.x semver minor to accommodate these changes. It looks like there are semver minor changes outside of http2, but if they scoped only to http2 we would be able to land as a semver-patch (as it is not stable in 8.x)

kjin commented:
@MylesBorins With the exception of adding the `ERR_TLS_RENEGOTIATE` error, do you know what else would count as semver-minor that is outside of the H2 domain?

MylesBorins commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/17592/

kjin commented:
Update #2: I've added commits that were landed between 10.9 and 10.11. There were some in particular that I opted to skip, that might need additional attention:

* d3ceaa1d41 (#22252): I changed tests related to timeouts. More info in https://github.com/nodejs/node/pull/22252#issuecomment-426784477.
* 2a88731ed7 (#22449): This PR has a lot of non-http2 stuff going on.
* 24675a4306 (#22956): This change makes use of a V8 API that's not available in Node 8. I use the closest approximation of the API (`StringOneByte` overload) instead. (The corresponding commit in this PR has a similar note.)
* d1826fed41 (#23047) This PR appears to fix an issue which might be the result of a PR that wasn't backported earlier (#18936 -- contains `do-not-backport-8.x` tag).

I believe as it is, this PR passes tests. However, there are a few things I want to get confirmation for from these folks:
* @apapirovski https://github.com/nodejs/node/pull/20084#issuecomment-422977248
* @jasnell https://github.com/nodejs/node/pull/22252#issuecomment-426784477 as well as -- does the third note in the list above make sense to you?
* @addaleax https://github.com/nodejs/node/pull/22850#issuecomment-422985422 as well as -- Does it make sense for me not to backport the second commit listed above?

addaleax commented:
@kjin It’s okay not to backport https://github.com/nodejs/node/commit/2a88731ed71fb17fc12ff159d59ca7f948cf1a70, yes

kjin commented:
Ping @nodejs/lts -- I think this is ready to move forward (let me know if you need anything from my end)

MylesBorins commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/17883/

BethGriggs commented:
The linter is failing, it seems to be this commit that is failing: 

```
commit 019e127df96e2a53fa5830446a7c206a4eb3d4e1
Author: James M Snell <jasnell@gmail.com>
Date:   Sun Sep 16 19:13:11 2018 -0700

    http2: add origin frame support
    
    v8.x Backport Note -- as V8 doesn't expose an overload of String::WriteOneByte
    in Node 8 that accepts an isolate argument, the Origins constructor has been
    changed to not accept an isolate.
    
    PR-URL: https://github.com/nodejs/node/pull/22956
    Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
```

kjin commented:
@BethGriggs Thanks, turns out I didn't have the MD linter installed. I've fixed the issue now.

BethGriggs commented:
Another CI: https://ci.nodejs.org/job/node-test-commit/22378 [passed]

MylesBorins commented:
@Flarna if you could open a new issue that documents how to recreate the observed behavior we can get to this ASAP

Flarna commented:
@MylesBorins  done: #24559
